tests/std/tests/P1222R4_flat_set/test.cpp

```diff
@@ -0,0 +1,213 @@
+    assert_container_requirements(s);
+    assert_reversible_container_requirements(s);
+
+    auto val_comp = s.value_comp();
+    auto begin_it = s.cbegin();
+    auto end_it   = s.cend();
+    assert(std::is_sorted(begin_it, end_it, val_comp));
+    if constexpr (!_Is_specialization_v<T, flat_multiset>) {
+        if (!s.empty()) {
+            auto it = begin_it;
+            while (++it != end_it) {
+                assert(val_comp(*(it - 1), *it));
+            }
+        }
+    }
+
+    if (s.size() != il.size() || !std::equal(s.begin(), s.end(), il.begin())) {
+        cout << "Expected: {";
+        for (auto&& e : il) {
+            cout << e << ", ";
+        }
+        cout << "}" << endl;
+        cout << "Got:      {";
+        for (auto&& e : s) {
+            cout << e << ", ";
+        }
+        cout << "}" << endl;
+        assert(false);
+    }
+}
+
+template <class T>
+void assert_basic() {
+    T s{3, 2, 2, 2, 1};
+    assert_all_requirements_and_equals(s, {1, 2, 3});
+
+    s.insert(43);
+    assert_all_requirements_and_equals(s, {1, 2, 3, 43});
+
+    int my_ints[] = {1, 2, 3, 4, 55};
+    s.insert_range(my_ints);
+    assert_all_requirements_and_equals(s, {1, 2, 3, 4, 43, 55});
+}
+
+template <class C>
+void test_constructors() {
+    using lt = std::less<int>;
+    using gt = std::greater<int>;
+
+    assert_all_requirements_and_equals(flat_set<int, lt, C>(), {});
+    assert_all_requirements_and_equals(flat_multiset<int, lt, C>(), {});
+    assert_all_requirements_and_equals(flat_set<int, lt, C>(C{3, 7, 1, 85, 222, 1}), {1, 3, 7, 85, 222});
+    assert_all_requirements_and_equals(flat_multiset<int, lt, C>(C{3, 7, 1, 85, 7, 222, 1}), {1, 1, 3, 7, 7, 85, 222});
+    assert_all_requirements_and_equals(flat_set<int, gt, C>(C{1, 2, 3, 3}, gt()), {3, 2, 1});
+    assert_all_requirements_and_equals(flat_multiset<int, gt, C>(C{1, 1, 2, 3}, gt()), {3, 2, 1, 1});
+    assert_all_requirements_and_equals(flat_set<int, gt, C>(sorted_unique, C{30000, 200, 1}, gt()), {30000, 200, 1});
+    assert_all_requirements_and_equals(flat_multiset<int, gt, C>(sorted_equivalent, C{3, 3, -1}, gt()), {3, 3, -1});
+    assert_all_requirements_and_equals(flat_set<int, gt, C>({30000, 200, 1}, gt()), {30000, 200, 1});
+    assert_all_requirements_and_equals(flat_multiset<int, gt, C>({3, 3, -1}, gt()), {3, 3, -1});
+    assert_all_requirements_and_equals(flat_set<int, gt, C>(sorted_unique, {30000, 200, 1}, gt()), {30000, 200, 1});
+    assert_all_requirements_and_equals(flat_multiset<int, gt, C>(sorted_equivalent, {3, 3, -1}, gt()), {3, 3, -1});
+
+    flat_set<int> a{};
+    a = {1, 7, 7, 7, 2, 100, -1};
+    assert_all_requirements_and_equals(a, {-1, 1, 2, 7, 100});
+    flat_multiset<int> b{};
+    b = {1, 7, 7, 7, 2, 100, -1};
+    assert_all_requirements_and_equals(b, {-1, 1, 2, 7, 7, 7, 100});
+}
+
+template <class T>
+void test_spaceship_operator() {
+    static constexpr bool multiset = _Is_specialization_v<T, flat_multiset>;

```

Comment:
We try to avoid having test identifiers collide with/shadow Standard identifiers. Since `multiset` is an STL type, it would be clearer to name this `is_multiset` or just `multi`.